cw_infrastructure/cw_infrastructure/validators.py

```diff
@@ -0,0 +1,278 @@
+
+
+def domain_name_validator(name, value):
+    """Validate a config option that should be a domain name"""
+    if not utils.is_domain_name(value):
+        error(name, "{} is not a valid domain name".format(value))
+        return ERROR
+    else:
+        return OK
+
+
+def ip_or_domain_name_validator(name, value):
+    """Validate a config option that should be a domain name or IP address"""
+    if not utils.is_ip_addr(value) and not utils.is_domain_name(value):
+        error(name, "{} is neither a valid IP address or domain name".format(value))
+        return ERROR
+    else:
+        return OK
+
+
+def resolvable_domain_name_validator(name, value):
+    """Validate a config option that should be a resolvable domain name"""
+    if not utils.is_domain_name(value):
+        error(name, "{} is not a valid domain name".format(value))
+        return ERROR
+
+    if utils.is_resolvable_domain_name(value):
+        return OK
+    else:
+        error(name, "{} is not resolvable".format(value))
+        return ERROR
+
+
+def sip_uri_validator(name, value):
+    """Validate a config option represents a valid SIP URI"""
+
+    match = re.match(r"^([a-z]+):(?:([^@])+@)?([^:;]*)(?::(\d+))?(?:;(.*))?$", value)
+
+    if not match:
+        error(name, "{} is not a valid SIP URI".format(value))
+        return ERROR
+
+    scheme = match.group(1)
+    host = match.group(3)
+    port = match.group(4)
+    params = match.group(5)
+
+    if scheme != "sip":
+        error(name, "{} is not a SIP URI".format(value))
+        return ERROR
+
+    if params:
+        pdict = {}
+
+        for param in params.split(';'):
+            pmatch = re.match(r"([^=]+)=(.*)", param)
+            if pmatch:
+                pdict[pmatch.group(1)] = pmatch.group(2)
+            else:
+                pdict[param] = True
+
+        params = pdict
+    else:
+        params = {}
+
+    # Check whether a transport parameter is provided and is valid
+    transport = None
+
+    if 'transport' in params:
+        if params['transport'].lower() not in ('udp', 'tcp'):
+            error(name, ("{} is not a valid SIP "
+                         "transport").format(params['transport']))
+            return ERROR
+
+        else:
+            transport = params['transport']
+
+    if utils.is_ip_addr(host):
+        return OK
+
+    elif not utils.is_domain_name(host):
+        error(name, ("{} is neither an IP address or a valid domain "
+                     "name".format(host)))
+        return ERROR
+
+    elif port:
+        if utils.is_resolvable_domain_name(host):
+            return OK
+
+        else:
+            error(name, "{} is not resolvable".format(host))
+            return ERROR
+
+    elif transport:
+        srv = '_sip._{}.{}'.format(params['transport'], host)
+
+        if utils.is_srv_resolvable(srv):
+            return OK
+        else:
+            error(name, "{} is not a valid SRV record".format(srv))
+            return ERROR
+    else:
+
```

Comment:
Unneeded blank line (probably already in my code).

Reply:
I've removed this.